kubernetes-client/src/main/java/io/fabric8/kubernetes/client/CustomResource.java

```diff
@@ -229,4 +253,97 @@ public T getStatus() {
+    
+    Object instantiate() throws Exception;
+    
+    /**
+     * No-op instantiator.
+     */
+    Instantiator NULL = () -> null;
+  }
+  
+  /**
+   * Returns the {@link Instantiator} instance associated with the type parameter associated with the specified index in the
+   * generic type definition. Records the result so that it's only done once per CustomResource implementation.
+   *
+   * @param genericTypeIndex the index of the parameter type we want to instantiate in the CustomResource definition, i.e. with
+   *                         the CustomResource<S,T> definition, {@code 0} corresponds to the {@code S} type (i.e. the Spec type)
+   *                         while {@code 1} corresponds to the {@code T} (i.e. Status) type.
+   * @return the {@link Instantiator} for the desired type
+   * @throws Exception if the generic type cannot be identified or instantiated
+   */
+  private Instantiator getInstantiator(int genericTypeIndex) throws Exception {
+    final String key = getKey(getClass(), genericTypeIndex);
+    Instantiator instantiator = instantiators.get(key);
+    if (instantiator == null) {
+      instantiator = Instantiator.NULL;
+      
+      // walk the type hierarchy until we reach CustomResource or a ParameterizedType
+      Type genericSuperclass = getClass().getGenericSuperclass();
+      String typeName = genericSuperclass.getTypeName();
+      while (!TYPE_NAME.equals(typeName) && !(genericSuperclass instanceof ParameterizedType)) {
+        genericSuperclass = ((Class) genericSuperclass).getGenericSuperclass();
+        typeName = genericSuperclass.getTypeName();
+      }
+  
+      // this works because CustomResource is an abstract class
+      if (genericSuperclass instanceof ParameterizedType) {
+        final Type[] types = ((ParameterizedType) genericSuperclass).getActualTypeArguments();
+        if (types.length != 2) {
+          throw new IllegalArgumentException(
+            "Automatic instantiation of Spec and Status only works for CustomResource implementations parameterized with both types, consider overriding initSpec and/or initStatus");
+        }
+        // get the associated class from the type name, if not Void
+        String className = types[genericTypeIndex].getTypeName();
+        if (!VOID_TYPE_NAME.equals(className)) {
+          Class<?> clazz = Class.forName(className);
+          if (clazz.isInterface() || Modifier.isAbstract(clazz.getModifiers())) {
+            throw new IllegalArgumentException(
+              "Cannot instantiate interface/abstract type " + className);
+          }
+          
+          // record the instantiator associated with the identified type
+          instantiator = () -> {
+            final Constructor<?> constructor;
+            try {
+              // get the no-arg (declared needed if implicit) constructor
+              constructor = clazz.getDeclaredConstructor();
+              constructor.setAccessible(true); // and make it accessible
```

Comment:
_Thought:_
This line worries me.
Can't think  of a case where the no-arg constructor will not be accessible. 
And in case it's not, we should probably desist in providing the empty instance since the user might have added some functionality to the alternate constructor. Invoking the hidden no-arg constructor will probably defeat the user's expectations and security assumptions.

Anyway, still not sure if this is a big deal. Everything else LGTM.

Reply:
I had to add this to make the most common case (a class that doesn't provide any constructor at all, i.e. one that relies on the implicit one) work so it's pretty much required. That surprised me as well.

Reply:
:flushed: that's really strange, not sure how Jackson deals with this then.